collector/src/test/java/com/google/cloud/trace/zipkin/ZipkinCollectorIntegrationTest.java

```diff
@@ -0,0 +1,143 @@
+import static java.lang.Math.min;
+import static java.util.Arrays.asList;
+import static java.util.Collections.unmodifiableList;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.hasItems;
+import static org.hamcrest.Matchers.hasSize;
+import static org.mockito.Mockito.mock;
+import static zipkin.TestObjects.LOTS_OF_SPANS;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = {StackdriverZipkinCollector.class, ZipkinCollectorIntegrationTest.TestConfiguration.class},
+    webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT,
+    properties = {ZIPKIN_CONFIG_NAMES, "zipkin.storage.stackdriver.project-id=test_project"})
+public class ZipkinCollectorIntegrationTest
+{
+  private static final Logger LOG = LoggerFactory.getLogger(ZipkinCollectorIntegrationTest.class);
+
+  //LOTS_OF_SPANS is too much! I need at most 30!
+  private static final List<Span> TEST_SPANS = unmodifiableList(asList(Arrays.copyOfRange(LOTS_OF_SPANS, 0, min(30, LOTS_OF_SPANS.length))));
+
+  private static int GRPC_PORT;
+
+  @Autowired
+  private TestRestTemplate restTemplate;
+  private Server server;
+  private Map<Long, Span> expectedSpans;
+
+  @BeforeClass
+  public static void beforeClass() throws UnknownHostException
+  {
+    GRPC_PORT = SocketUtils.findAvailableTcpPort();
+    System.setProperty("zipkin.storage.stackdriver.api-host", "localhost:" + GRPC_PORT);
+  }
+
+  @Before
+  public void before() throws IOException
+  {
+    this.server = NettyServerBuilder.forPort(GRPC_PORT).addService(new MockTraceService()).build().start();
+    LOG.info("Started MOCK grpc server on 'localhost:{}'", GRPC_PORT);
+  }
+
+  @After
+  public void after()
+  {
+    this.server.shutdownNow();
+  }
+
+  @Test
+  public void traceSinkGetsCalled() throws InterruptedException
+  {
+    this.expectedSpans = TEST_SPANS.stream().collect(Collectors.toConcurrentMap(s -> s.id, Function.identity()));
+    assertThat("Expected to process at least one span", this.expectedSpans.keySet(), hasSize(greaterThan(0)));
+
+    for (List<Span> partition : Lists.partition(TEST_SPANS, 3))
+    {
+      this.restTemplate.postForLocation("/api/v1/spans", Codec.JSON.writeSpans(partition));
+    }
+
+    assertThat("There are some unprocessed spans left", this.expectedSpans.keySet(), hasSize(equalTo(0)));
+  }
+
+  @Configuration
+  public static class TestConfiguration
+  {
+    @Bean(name = "stackdriverExecutor")
+    @Primary
+    public Executor syncExecutor()
+    {
+      //executes each task synchronously
+      return new SyncTaskExecutor();
+    }
+
+    @Bean(name = "googleCredentials")
+    @Primary
+    public Credentials mockGoogleCredentials() throws IOException
+    {
+      return mock(GoogleCredentials.class);
+    }
+  }
+
+  class MockTraceService extends TraceServiceGrpc.TraceServiceImplBase
```

Comment:
cool :)

not saying you need to do this, but one option is to extract this out as a junit rule, and expose the traces received as a collection. It would make the test itself easier to read since the infrastructure around the fake server would be somewhere else.

ex.

```java
@Rule public StackDriverTraceRule traceService = new StackDriverTraceRule(PORT); // specify the port or allow something to look it up.


// then later 
 @Test
public void tracesSunk() throws InterruptedException {
  // do something to produce traces

  // then either use a blocking queue internally in your rule and take a patch requests explicitly like this.
  assertThat(traceService.takePatchRequest()) // using assertj this works..
     .extracting(r -> getTraces().getTracesList())
     .flatExtracting(t -> t.getSpansList())
     .extracting(Span::getSpanId)
     .hasCount(3); // or containsExactly(myIds)
  
  // or just sink them to a local collection and interact directly
  assertThat(traceService.traceIds())
     .containsExactly(...)

}

```

https://github.com/openzipkin/zipkin/blob/master/zipkin-junit/src/main/java/zipkin/junit/ZipkinRule.java